text/3366-diagnostic-attribute-namespace.md

```diff
@@ -0,0 +1,233 @@
+
+Attributes in this namespace are generally expected to be formed like:
+```rust
+#[diagnostic::attribute(option)]
+```
+where several `option` entries can appear in the same attribute. `option` is expected to be a valid attribute argument in this position. 
+
+Any attribute in this namespace may:
+
+* Hint the compiler to emit a specific diagnostic message in a specific situation
+* Only affect the messages emitted by a compiler 
+
+Any attribute in this namespace is not allowed to:
+
+* Change the result of the compilation, which means applying such an attribute should never cause a compilation error as long as they are syntactically valid
+* Pass-through information from the source of the diagnostic in a way that users can rely on. E.g. Such an attribute should not allow users to keep the compilation successful and dump information about `extern` blocks to generate C header files
+
+The compiler is allowed to:
+
+* Ignore the hints provided by:
+    + A specific attribute
+    + A specific option 
+* Change the support for a specific attribute or option at any time
+
+The compiler must not:
+
+* Change the semantic of an attribute or option
+* Emit an hard error on malformed attribute
+
+The compiler should:
+
+* Emit implement a warn-by-default lint for unrecognised attributes or options 
+
+Adding a new attribute or option to the `#[diagnostic]` namespace is for now a decision of the language team. The language team can delegate these decisions partially or completely to a different team without requiring a new RFC. 
+
+## The `#[diagnostic::on_unimplemented]` attribute
+
+This section describes the syntax of the `on_unimplemented` attribute and additionally how it is supposed to work. The specification of this attribute is partially provided as example and motivation for the `#[diagnostic]` attribute namespace. In addition it is provided to give this RFC a concrete use, such that we not only define an empty attribute namespace.
+
+```rust
+#[diagnostic::on_unimplemented(
+    message="message",
+    label="label",
+    note="note",
+)]
+trait MyIterator<A> {
+    fn next(&mut self) -> A;
+}
+```
+
+
+Each of the options `message`, `label` and `note` are optional. They are separated by comma. The trailing comma is optional. Specifying any of these options hints the compiler to replace the normally emitted part of the error message with the provided string. At least one of these options needs to exist. Each option can appear at most once. These options can include type information for the `Self` type or any generic type by using `{Self}` or `{A}` (where `A` refers to the generic type name in the definition). These placeholders are replaced by the compiler with the actual type name.
```

Comment:
I understand the idea of *replacing* the message and the label, in the sense that I think there is always a message, and there is one label (potentially *at most one* label? Can the label be omitted? Is this inserted in its stead in that case?). So replacing each of those is the sensible option here.

However, a given diagnostic can have zero, one, or more than one notes supplied, right?. What is *replaced* for the "more than one" case, here? And would it be better to just accumulate this note onto the set, or are we indeed better off sticking with replacement here?

Reply:
What the best approach is depends entirely on the situation, sadly (which tells me it should be configurable somehow). Currently `rustc_on_unimplemented` will replace these but add a note with the original message.

I can see for example errors that would be very verbose where the underlying problem is very simple, so as a crate author I'd like to be able to almost entirely clear the diagnostic and clearly state "this is wrong and this is how you fix it", avoiding that messaging being lost in the noise. That being said there would have to be a consideration of whether this kind of "obscuring" of the emitted error could constitute a problem if misused. All this being said, I believe these are specific questions to be had around the stabilization of `#[diagnostic::on_unimplemented]` and not the acceptance of `#[diagnostic::*]` as a construct.